engine/runtime/src/main/java/org/enso/interpreter/runtime/data/text/Text.java

```diff
@@ -1,36 +1,60 @@
 package org.enso.interpreter.runtime.data.text;
 
+import com.ibm.icu.text.BreakIterator;
 import com.oracle.truffle.api.CompilerDirectives;
 import com.oracle.truffle.api.dsl.Cached;
 import com.oracle.truffle.api.interop.InteropLibrary;
 import com.oracle.truffle.api.interop.TruffleObject;
 import com.oracle.truffle.api.library.CachedLibrary;
 import com.oracle.truffle.api.library.ExportLibrary;
 import com.oracle.truffle.api.library.ExportMessage;
+import java.util.ArrayDeque;
+import java.util.Deque;
 import org.enso.interpreter.node.expression.builtin.text.util.ToJavaStringNode;
 import org.enso.interpreter.runtime.Context;
 import org.enso.interpreter.runtime.data.Type;
 import org.enso.interpreter.runtime.library.dispatch.TypesLibrary;
 
 import java.util.concurrent.locks.Lock;
 import java.util.concurrent.locks.ReentrantLock;
+import org.enso.interpreter.dsl.Builtin;
 
 /** The main runtime type for Enso's Text. */
 @ExportLibrary(InteropLibrary.class)
 @ExportLibrary(TypesLibrary.class)
 public final class Text implements TruffleObject {
+  private static final Lock LOCK = new ReentrantLock();
   private volatile Object contents;
-  private volatile boolean isFlat;
-  private static final Lock lock = new ReentrantLock();
+  private volatile int length = -1;
 
   private Text(String string) {
     this.contents = string;
-    this.isFlat = true;
   }
 
   private Text(ConcatRope contents) {
     this.contents = contents;
-    this.isFlat = false;
+  }
+
+  @Builtin.Method(description = """
+  Computes the number of characters in the text.
+
+    ! What is a Character?
+      A character is defined as an Extended Grapheme Cluster, see Unicode
+      Standard Annex 29. This is the smallest unit that still has semantic
+      meaning in most text-processing applications.
+
+    > Example
+      Getting the length of the string "건반(Korean)".
+
+          "건반(Korean)".length
+   """)
+  public long length() {
+    int l = length;
+    if (l == -1) {
+      l = computeLength();
+      length = l;
+    }
+    return l;
```

Comment:
I think I don't understand why this helper variable is ever needed?
Seems to me that the following code will work exactly the same and is much more readable:
```suggestion
    if (length == -1) {
      length = computeLength();
    }
    return length;
```

Am I missing something here?

Reply:
Is this the fact that `length` is volatile so its reads are more costly and we do not want to read it twice? Maybe a short comment then would be helpful - as it is easy to miss this fact and get confused.
```suggestion
    // Storing the length value to avoid doubling the cost of the volatile variable read.
    int l = length;
    if (l == -1) {
      l = computeLength();
      length = l;
    }
    return l;
```

Reply:
Right. `length` is volatile, so I want to read it just once per method invocation. I was afraid that your first version would read the field twice.

> short comment then would be helpful

Rather than a comment, I should have given the variable more descriptive name - `localLength`? 

Anyway rather than guessing, I tried the `StringBenchmarks` with
```
enso$ git diff
diff --git a/engine/runtime/src/main/java/org/enso/interpreter/runtime/data/text/Text.java b/engine/runtime/src/main/java/org/enso/interpreter/runtime/data/text/Text.java
index 43ade9540..1b846daba 100644
--- a/engine/runtime/src/main/java/org/enso/interpreter/runtime/data/text/Text.java
+++ b/engine/runtime/src/main/java/org/enso/interpreter/runtime/data/text/Text.java
@@ -49,12 +49,10 @@ public final class Text implements TruffleObject {
           "건반(Korean)".length
    """)
   public long length() {
-    int l = length;
-    if (l == -1) {
-      l = computeLength();
-      length = l;
+    if (length == -1) {
+      length = computeLength();
     }
-    return l;
+    return length;
   }
 
   /**
```
it is as fast the version with local variable. Interesting. Good to know for next time.

Reply:
Interesting indeed, thanks for checking it.